Mapping/Loader/SerializerMetadataLoader.php

```diff
@@ -0,0 +1,184 @@
+        }
+
+        if ($this->hasGroups($groups, $denormalizationGroups)) {
+            $attributeMetadata = $attributeMetadata->withWritable(true);
+            $classMetadata = $classMetadata->withAttributeMetadata($attributeName, $attributeMetadata);
+        }
+
+        return [$classMetadata, $attributeMetadata];
+    }
+
+    /**
+     * Checks if an attribute has the passed groups.
+     *
+     * @param array      $groups
+     * @param array|null $currentGroups
+     *
+     * @return bool
+     */
+    private function hasGroups(array $groups, array $currentGroups = null)
+    {
+        return null !== $currentGroups && 0 < count(array_intersect($groups, $currentGroups));
+    }
+
+    /**
+     * Populates normalization and denormalization links.
+     *
+     * @param ClassMetadataInterface     $classMetadata
+     * @param string                     $attributeName
+     * @param AttributeMetadataInterface $attributeMetadata
+     * @param array|null                 $normalizationGroups
+     * @param array|null                 $denormalizationGroups
+     *
+     * @return ClassMetadataInterface
+     */
+    private function populateNormalizationLinks(
+        ClassMetadataInterface $classMetadata,
+        $attributeName,
+        AttributeMetadataInterface $attributeMetadata,
+        array $normalizationGroups = null,
+        array $denormalizationGroups = null
+    ) {
+        if (
+            !$classMetadata->hasAttributeMetadata($attributeName) ||
+            !$attributeMetadata->isLink() ||
+            ($attributeMetadata->isNormalizationLink() && $attributeMetadata->isDenormalizationLink())
+        ) {
+            return $classMetadata;
+        }
+
+        $relationSerializerMetadata = $this->serializerClassMetadataFactory->getMetadataFor($attributeMetadata->getLinkClass());
+        if (!$relationSerializerMetadata) {
+            $attributeMetadata = $attributeMetadata->withNormalizationLink(true)->withDenormalizationLink(true);
+
+            return $classMetadata->withAttributeMetadata($attributeName, $attributeMetadata);
+        }
+
+        foreach ($relationSerializerMetadata->getAttributesMetadata() as $serializerAttributeMetadata) {
+            $serializerAttributeGroups = $serializerAttributeMetadata->getGroups();
+
+            if (null !== $normalizationGroups && 1 <= count(array_intersect($normalizationGroups, $serializerAttributeGroups))) {
+                $normalizationLink = false;
+            }
+
+            if (null !== $denormalizationGroups && 1 <= count(array_intersect($denormalizationGroups, $serializerAttributeGroups))) {
+                $denormalizationLink = false;
+            }
+
+            if (isset($normalizationLink) && isset($denormalizationLink)) {
+                $classMetadata = $classMetadata->withAttributeMetadata($attributeName, $attributeMetadata);
+
+                continue;
```

Comment:
As it's the last statement of the `foreach` content... it'll `continue` anyway :)